[Merged by Bors] - Remove ModelGen

devmotion opened the pull request:
This PR removes `ModelGen` completely. The main motivation for it were the issues that @itsdfish and I experienced when working with multiple processes. The following MWE
```julia
using Distributed
addprocs(4)
@everywhere using DynamicPPL

@everywhere @model model() = begin end

pmap(x -> model(), 1:4)
```
fails intermittently

> if not all of these [`@model`] evaluations generate same evaluator and generator functions of the same name (i.e., these var"###evaluator#253" and var"###generator#254" functions). I assume one could maybe provoke the error by defining another model first on the main process before calling @everywhere @model ....

(copied from the discussion on Slack)

With the changes in this PR, `@model model() = ...` generates only a single function `model` on all workers, and hence there are no issues anymore with the names of the generators and evaluators. The evaluator is created as a closure inside of `model`, which can be serialized and deserialized properly by Julia. So far I haven't been able to reproduce the issues above with this PR.

The only user-facing change of the removal of `ModelGen` (apart from that one never has to construct it, which simplifies the docs and the example @denainjs asked about) is that the `logprob` macro now requires to specify `model = m` where `m = model()` instead of `model = model` (since that's just a regular function from which the default arguments etc of the resulting model can't be extracted). It feels slightly weird that the evaluation is not based "exactly" on the specified `Model` instance but that the other parts of `logprob` modify it (which was the reason I guess for using the model generator before here), but on the other hand this weird behaviour already exists when specifying `logprob` with a `Chains` object. (BTW I'm not sure if we should actually use string macros here, maybe regular functions would be nicer.)

Additionally, I assume (though haven't tested it) that getting rid of the separate evaluator and generator functions will not only simplify serialization and deserialization when working with multiple processes but also when saving models and chains (see e.g. https://github.com/TuringLang/Turing.jl/issues/1091).

itsdfish commented:
Thank you for looking into this! It looks like this change will simplify several things.

yebai commented on src/context_implementations.jl:
```diff
@@ -392,7 +392,7 @@ function dot_tilde(ctx::LikelihoodContext, sampler, right, left, vi)
 end
 function dot_tilde(ctx::MiniBatchContext, sampler, right, left, vi)
-    return ctx.loglike_scalar * dot_tilde(ctx.ctx, sampler, right, left, left, vi)
+    return ctx.loglike_scalar * dot_tilde(ctx.ctx, sampler, right, left, vi)
```
It's strange that these previous merged PRs are still shown as new changes. Perhaps rebase this PR with the `dev` branch?

phipsgabler commented on src/model.jl:
```diff
@@ -1,112 +1,70 @@
 
-julia> Model{(:y,)}(f, (x = 1.0, y = 2.0))
-Model{typeof(f),(:y,),(:x, :y),Tuple{Float64,Float64}}((x = 1.0, y = 2.0))
+julia> Model{(:y,)}(f, (x = 1.0, y = 2.0), (x = 42,))
```
Maybe add a comment here that this is the "overwrite missings" constructor.

phipsgabler commented on src/model.jl:
```diff
@@ -184,27 +142,19 @@ end
-
 """
     logjoint(model::Model, varinfo::AbstractVarInfo)
 
```
There are some worrysome tabs in lines 167, 179, and 191.  (Github doesn't let me comment there directly.)

devmotion replied:
Oh :open_mouth: I'll fix it.

phipsgabler commented on src/compiler.jl:
```diff
@@ -340,26 +340,26 @@ function build_output(model_info)
-            $unwrap_data_expr
-            $main_body
+        $(Base).@__doc__ function $model($(args...))
+            $evaluator = let
```
Is the `let` really needed here, or just cautiously used?

devmotion replied:
I haven't tested it, but the main motivation was to ensure that none of the arguments `args` is captured by the evaluator.

phipsgabler replied:
Ah, but I think that doesn't work with an empty `let`:

```julia
julia> function f1(x)
           g = function()
               x += 2
               return x
           end
           return x, g(), x
       end
f1 (generic function with 1 method)

julia> f1(1)
(1, 3, 3)
```

is the same as

```julia
julia> function f2(x)
           g = let
           function()
               x += 2
               return x
           end
           end
           return x, g(), x
       end
f2 (generic function with 2 methods)

julia> f2(1)
(1, 3, 3)
```

An empty definition prevents access completely:

```julia
julia> function f3(x)
           g = let x
           function()
               x += 2
               return x
           end
           end
           return x, g(), x
       end
f3 (generic function with 1 method)

julia> f3(1)
ERROR: UndefVarError: #45#x not defined
Stacktrace:
 [1] (::var"#23#24")() at ./REPL[27]:4
 [2] f3(::Int64) at ./REPL[27]:8
 [3] top-level scope at REPL[28]:1
```

Could we just turn `unwrap_data_expr` into a `let` around `main_body`, but within the closure, instead?

devmotion replied:
Hmm, actually somehow I'm not a big fan of `unwrap_data_expr` and I'm wondering if the `let` block inside of the function could impact performance and the output of the compiler in any way. Maybe it would be more natural to define
```julia
$(Base).@__doc__ function $model($(args...))
            $evaluator = function (
                _rng::$(Random.AbstractRNG),
                _model::$(DynamicPPL.Model),
                _varinfo::$(DynamicPPL.AbstractVarInfo),
                _sampler::$(DynamicPPL.AbstractSampler),
                _context::$(DynamicPPL.AbstractContext),
                $(arg_syms...)
            )
                $main_body
            end
            return $(DynamicPPL.Model)($evaluator, $args_nt, $defaults_nt)
        end
```
(in this case none of the arguments of `$model` can be captured by the evaluator), and then just call the evaluator always with the arguments replaced with their "matching values". For this one could use a generated function (maybe not needed, would just avoid splatting) such as
```julia
@generated function _evaluate(rng, model::Model{_F,argnames}, varinfo, sampler, context) where {_F,argnames}
    unwrap_args = [:($matchingvalue(sampler, varinfo, model.args.$var)) for var in argnames]
    return :(model.f(rng, model, varinfo, sampler, context, $(unwrap_args...)))
end
```

However, at a second glance maybe it's not actually simplifying anything here...

phipsgabler replied:
So, a function barrier for the arguments... sounds reasonable. :+1:

devmotion replied:
Yes, that was my idea instead of the `let` block. Ideally, though, I guess we would want to add the `where` clauses and use `::Type{T}` instead of `T` there to ensure that the compiler specializes on the type arguments.

phipsgabler replied:
Oh absolutely.  Can the `:whereparams` part of the model info be used directly?  

If only the MacroTools refactoring PR was finished yet...

devmotion replied:
Shouldn't be too difficult, just have to make the keyword argument regular arguments, add our arguments, replace the function body, and then use `MacroTools.combinedef`, similar to https://github.com/TuringLang/DynamicPPL.jl/pull/109.

devmotion replied:
OK, was not too difficult. I ended up using ExprTools.combinedef instead of MacroTools.combinedef since the latter could not handle functions without names in the latest release (AFAICT it should work with master though).

Along the same lines, we might want to consider something like https://github.com/TuringLang/DynamicPPL.jl/pull/109 to keep the user-provided function signature for the model definition (which is separate from the evaluator definition and this PR).

yebai commented on Project.toml:
```diff
@@ -7,6 +7,7 @@ version = "0.8.1"
 AbstractMCMC = "80f14c24-f653-4e6a-9b94-39d6b0f70001"
 Bijectors = "76274a88-744f-5084-9051-94815aaf08c4"
 Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
+ExprTools = "e2ba6199-217a-4e67-a87a-7c52f15ade04"
```
`MacroTools` has reduced dependency footprint recently. There are also some discussions about merging `ExprTools` into `MacroTools`, see https://github.com/MikeInnes/MacroTools.jl/pull/132#issuecomment-633683482

The changes in https://github.com/TuringLang/DynamicPPL.jl/pull/121 might be no longer necessary.

devmotion replied:
I agree that it's annoying to depend on both packages. I didn't follow the development in MacroTools lately, nice to here if dependency (and probably loading times) could be improved! It seems

> I ended up using ExprTools.combinedef instead of MacroTools.combinedef since the latter could not handle functions without names in the latest release (AFAICT it should work with master though).

I'll try to work around this limitation, so that ExprTools is not needed.

devmotion replied:
I removed ExprTools again and simplified the extraction of the arguments and their default values as well.

phipsgabler commented:
> > I'm not sure if we should actually use string macros here, maybe regular functions would be nicer.)
> 
> I agree the string macro is a bit strange and would be nice to consider a regular function replacement.

I have never used it, but also wondered about that choice.  We could use some kind of formula objects, like [GLM does](https://juliastats.org/GLM.jl/dev/examples/), although "formula" is a strange name and we shouldn't reuse the tilde here.  And unfortunately, `|` binds stronger than assignment and comma, so `@prob(x = 1 | y = 2)` parses as `(= x (= (| 1 y) 2))`.

On the other hand, wouldn't something like

```julia
density(model, s = 1.0, m = 1.0, given = (;x = 1.0), chain = c)
```

suffice to analyze which variables go where and perform the according call?

devmotion commented:
Regarding the precedence issue, one could use, e.g., `@logprob (x = 1) | (y = 3, z = 24)` which seems to parse correctly.

In fact that would be trivial to implement it seems, since it could just be lowered to the `logprob` function in https://github.com/TuringLang/DynamicPPL.jl/blob/275ccc8791be7d5dc4eb74b6e0c19de247f56d74/src/prob_macro.jl#L24. It seems the `logprob"..."` macro just wraps the LHS and RHS of `|` as tuples and then calls `logprob`, so it's already like the `density` suggestion with the model and chain being part of the RHS.

IMO it might be more natural to keep the `model` and the `chain` that is used in these computations separate from the LHS and RHS of `|` since it can only be part of the RHS anyways. Similar to the `density` suggestion, one could just require the `model` and/or the `chain` to be the first argument of `density`/`logprob`. That would also avoid issues if a variable is called `chain` or `model`. Maybe it would also resolve the current precedence issues (at least there were some discussions and issues a while ago IIRC) about what to do when both a chain with a saved model and a model are specified since we could define the function signature and provided dispatches at will. In the same way the `@logprob` macro could be extended with additional arguments that are not part of the tuples.

phipsgabler commented:
> Regarding the precedence issue, one could use, e.g., `@logprob (x = 1) | (y = 3, z = 24)` which seems to parse correctly.

Sure, but that would 1) force the user to use parentheses in a place where "semantically" the form without parentheses should be equally valid, and 2) will hit people with the subtle distinction between `@macro (foo)` and `@macro(foo)`.  That's why I didn't suggest that form.

Now I get why this was made a string macro :D

> In fact that would be trivial to implement it seems, since it could just be lowered to the `logprob` function [...]. It seems the `logprob"..."` macro just wraps the LHS and RHS of `|` as tuples and then calls `logprob`, so it's already like the `density` suggestion with the model and chain being part of the RHS.
> 
> IMO it might be more natural to keep the `model` and the `chain` that is used in these computations separate from the LHS and RHS of `|` since it can only be part of the RHS anyways. Similar to the `density` suggestion, one could just require the `model` and/or the `chain` to be the first argument of `density`/`logprob`. 

Exactly my thought.

New suggestion for sugar: using `;` for conditioning, make `@pr` construct a "probability formula", somewhat like the mathematical notation `Pr(p1(x), p2(y); z)`, where we deal with predicates of `x` and `y`, and pass this to `logprob` and `logdensity` functions, just that we always have to be explicit about names:

```julia
logprob(model, @pr(x == 1; alpha = 2))
logprob(chain, @pr y > 0)
```

(And yes, I realize that this suffers from the same `@macro (foo)` problem :( )

One thing I'm just not sure about is how rigidly we should separate densities and probabilities.  That's why I called the function `density`.  I'd rather not further propagate the confusion many people have distinguishing between them (and likelihoods) by calling them the same.

devmotion commented:
I guess the main motivation for a macro over just having a function `logdensity(model, ::Tuple, ::Tuple)` would be that it might be more convenient for specifying the predicates?

devmotion commented:
BTW (since I just worked with it) the macro syntax with brackets and `|` is also used in the pattern matching in MacroTools: https://mikeinnes.github.io/MacroTools.jl/stable/pattern-matching/#Unions-1

devmotion commented:
Ah I just noticed that the latest commits broke the MWE above - one really has to use anonymous functions it seems. Unfortunately that means that currently we either have to add ExprTools as additional dependency or reimplement the logic in MacroTools master (I asked about a new release but I don't know how soon there'll be one). I guess it would be fine to copy the parts of MacroTools that we need right now and remove them if a new release is available?

yebai commented:
Sounds good.

devmotion commented:
This is ready for a second round of review @yebai @phipsgabler 

I removed ExprTools and copied the logic for combining anonymous functions from MacroTools (hopefully there's a release soon, then it can be removed again; I would also prefer actually if they merge ExprTools into MacroTools). I also simplified the code for extracting the arguments and their default values and used `MacroTools.@match` for increasing the readability when parsing expressions. If we stick with MacroTools for now, https://github.com/TuringLang/DynamicPPL.jl/pull/121 might not be needed anymore (or only in a reduced form).

phipsgabler commented:
Very nice!  I couldn't find any flaws in reading the code.  `build_model_info` got really slim, I agree that #121 might have become superfluous with this.

yebai commented:
bors r+

yebai commented:
bors r+
